### PR TITLE
[WIP]feat: add provider code

### PR DESCRIPTION
**Pull request by L2ncE**
Under continuous dev.

### CODE REVIEW

File: provider/service_config.go [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+			panic(perrors.New(fmt.Sprintf("Get tcp port error, err is {%v}", err)))
+		}
+		defer tcp.Close()
+		ports.PushBack(strings.Split(tcp.Addr().String(), ":")[1])
+	}
+	return ports
+}
+
+// ExportMethod exports the service
+func (s *ServiceConfig) ExportMethod() error {
```
Comment (DMwangnima) [inline provider/service_config.go:210]: There should be Export and ExportWithInfo like Refer and ReferWithInfo.

**Review by DMwangnima — CHANGES_REQUESTED** [review]

File: common/constant/key.go [hunk 2]
```diff
@@ -416,3 +416,10 @@ const (
 	MetricsMetadata = "dubbo.metrics.metadata"
 	MetricApp       = "dubbo.metrics.app"
 )
+
+// default meta cache config
+const (
+	DefaultMetaCacheName = "dubbo.meta"
+	DefaultMetaFileName  = "dubbo.metadata"
+	DefaultEntrySize     = 100
+)
```
Comment (justxuewei) [inline common/constant/key.go:425]: These lines of code should not belong to this patch. Please pull from the latest main branch and rebase onto it.

File: provider/provider.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+type MethodInfo struct {
+	Name           string
+	Type           string
+	ReqInitFunc    func() interface{}
+	StreamInitFunc func(baseStream interface{}) interface{}
+	MethodFunc     func(ctx context.Context, args []interface{}, handler interface{}) (interface{}, error)
+	Meta           map[string]interface{}
+}
+
+func NewProvider(opts ...ServiceOption) (*Provider, error) {
```
[the remark this answers is not in the document]
Comment (L2ncE) [reply provider/provider.go]: done

File: common/constant/key.go (hunk 2 above)
Comment (L2ncE) [reply common/constant/key.go]: done
